tick/stateful_expr_benchmarks_test.go

```diff
@@ -0,0 +1,293 @@
+	currentValue := initialValue
+	maximumValue := float64(40)
+
+	var result bool
+	for i := 0; i < b.N; i++ {
+		b.StopTimer()
+
+		currentValue += float64(1)
+		if currentValue > maximumValue {
+			currentValue = initialValue
+		}
+
+		scope.Set("value", currentValue)
+
+		b.StartTimer()
+
+		result, err := se.EvalBool(scope)
+		if err != nil || !result {
+			v, _ := scope.Get("value")
+			b.Errorf("Failed to evaluate: error=%v, result=%t, value=%v, init=%v, maximum=%v", err, result, v, initialValue, maximumValue)
+		}
+	}
+
+	evalBoolResult = result
+}
+
+func Benchmark_EvalBool_OneOperator_ReferenceNodeInt64_ReferenceNodeInt64(b *testing.B) {
+
+	scope := tick.NewScope()
+	scope.Set("l", int64(20))
+	scope.Set("r", int64(10))
+
+	benchmarkEvalBool(b, scope, &tick.BinaryNode{
+		Operator: tick.TokenGreater,
+		Left: &tick.ReferenceNode{
+			Reference: "l",
+		},
+		Right: &tick.ReferenceNode{
+			Reference: "r",
+		},
+	})
+}
+
+func Benchmark_EvalBool_OneOperatorWith11ScopeItem_ReferenceNodeInt64_NumberInt64(b *testing.B) {
+
+	scope := tick.NewScope()
+	scope.Set("value", int64(20))
+	for i := 0; i < 10; i++ {
+		scope.Set(fmt.Sprintf("value_%v", i), int64(i))
+	}
+
+	benchmarkEvalBool(b, scope, &tick.BinaryNode{
+		Operator: tick.TokenGreater,
+		Left: &tick.ReferenceNode{
+			Reference: "value",
+		},
+		Right: &tick.NumberNode{
+			IsInt: true,
+			Int64: int64(10),
+		},
+	})
+}
+
+func Benchmark_EvalBool_OneOperatorValueChanges_ReferenceNodeInt64_NumberInt64(b *testing.B) {
+
+	scope := tick.NewScope()
+	initialValue := int64(20)
+
+	scope.Set("value", initialValue)
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	se := tick.NewStatefulExpr(&tick.BinaryNode{
+		Operator: tick.TokenGreater,
+		Left: &tick.ReferenceNode{
+			Reference: "value",
+		},
+		Right: &tick.NumberNode{
+			IsInt: true,
+			Int64: int64(10),
+		},
+	})
+
+	// We have maximum value because we want to limit the maximum number in
+	// the reference node so we don't get too much big numbers and the benchmark suite will increase our iterations number (b.N)
+	currentValue := initialValue
+	maximumValue := int64(40)
+
+	var result bool
+	for i := 0; i < b.N; i++ {
+		b.StopTimer()
+
+		currentValue += int64(1)
+		if currentValue > maximumValue {
+			currentValue = initialValue
+		}
+
+		scope.Set("value", currentValue)
+
+		b.StartTimer()
+
+		result, err := se.EvalBool(scope)
+		if err != nil || !result {
+			v, _ := scope.Get("value")
+			b.Errorf("Failed to evaluate: error=%v, result=%t, value=%v, init=%v, maximum=%v", err, result, v, initialValue, maximumValue)
+		}
+	}
+
+	evalBoolResult = result
+}
+
+var evalBoolResult bool
+
+func benchmarkEvalBool(b *testing.B, scope *tick.Scope, node tick.Node) {
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	se := tick.NewStatefulExpr(node)
+	var result bool
+
+	for i := 0; i < b.N; i++ {
+		result, err := se.EvalBool(scope)
```

Comment:
This `result` is hiding the func level `result` var because of the `:=` operator. Declare a `var err error` outside the loop as well so you can use the `=` operator.